eyelid-match-ops/src/encrypted.rs

```diff
@@ -1 +1,210 @@
+            // (t - d) / 2t <= x%
+            #[allow(clippy::cast_possible_wrap)]
+            if (t - d) * (C::EyeConf::MATCH_DENOMINATOR as i64)
+                <= 2 * t * (C::EyeConf::MATCH_NUMERATOR as i64)
+            {
+                return Ok(true);
+            }
+        }
+
+        Ok(false)
+    }
+
+    /// Similarly to function `accumulate_inner_products`, but return a list containing the products, such that
+    /// we can extract inner products later.
+    fn accumulate_inner_products(
+        ctx: Yashe<C::PlainConf>,
+        private_key: PrivateKey<C::PlainConf>,
+        a_polys: &[Ciphertext<C::PlainConf>],
+        b_polys: &[Ciphertext<C::PlainConf>],
+    ) -> Result<Vec<i64>, MatchError>
+    where
+        BigUint: From<<C::PlainConf as PolyConf>::Coeff>,
+    {
+        let mut counts = vec![0; C::EyeConf::ROTATION_COMPARISONS];
+        // compute T/2 as a big int
+        let t_div_2 = BigInt::from(C::PlainConf::T / 2);
+
+        for (a, b) in a_polys.iter().zip_eq(b_polys.iter()) {
+            // Multiply the encrypted polynomials, which will yield encrypted inner products
+            // by the homomorphic property of the scheme.
+            let product = ctx.ciphertext_mul(a.clone(), b.clone());
+            // Decrypt to get the inner products.
+            let decrypted_product = ctx.decrypt_mul(product, &private_key);
+
+            // TODO: make the comparisons private
+            // Extract the inner products from particular coefficients.
+            // Left-most rotation:              sδ - (v - u) - 1
+            // Right-most rotation (inclusive): sδ - 1
+            let block_counts = decrypted_product
+                .m
+                .iter()
+                .skip(C::ROWS_PER_BLOCK * C::NUM_COLS_AND_PADS - C::EyeConf::ROTATION_COMPARISONS)
+                .take(C::EyeConf::ROTATION_COMPARISONS)
+                .map(|c| {
+                    let coeff_res = C::PlainConf::coeff_as_big_int(*c);
+                    // When the coefficient is negative, we need to convert it to work modulo T.
+                    // Concretely, we temporarily negate the coefficient in order to get a small value
+                    // (since negative elements modulo Q are big and can't be converted to i64), then we
+                    // negate again to return the output.
+                    if coeff_res > t_div_2 {
+                        let result = i64::try_from(BigUint::from(C::PlainConf::big_int_as_coeff(
+                            C::PlainConf::T - coeff_res,
+                        )))
+                        .expect("Could not convert a negative element to i64");
+                        Ok(-result)
+                    } else {
+                        let result =
+                            i64::try_from(BigUint::from(C::PlainConf::big_int_as_coeff(coeff_res)))
+                                .expect("Could not convert a positive from big int to i64");
```

Comment:
Let's return a new `MatchError` variant here, rather than panicking using `expect()`?

Reply:
This can become a TODO comment.